In [22]:
#Imports
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
#load csv file
df_churndata_modded = pd.read_csv('TelcoCustomerChurn[MODIFIED].csv')
df_churndata_modded.head()

Unnamed: 0  customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0           0  7590-VHVEG  Female              0     Yes         No       1   
1           1  5575-GNVDE    Male              0      No         No      34   
2           2  3668-QPYBK    Male              0      No         No       2   
3           3  7795-CFOCW    Male              0      No         No      45   
4           4  9237-HQITU  Female              0      No         No       2   

  PhoneService     MultipleLines InternetService  ... DeviceProtection  \
0           No  No phone service             DSL  ...               No   
1          Yes                No             DSL  ...              Yes   
2          Yes                No             DSL  ...               No   
3           No  No phone service             DSL  ...              Yes   
4          Yes                No     Fiber optic  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges TotalCharges  Churn  
0           Electronic check          29.85        29.85      0  
1               Mailed check          56.95       1889.5      0  
2               Mailed check          53.85       108.15      1  
3  Bank transfer (automatic)          42.30      1840.75      0  
4           Electronic check          70.70       151.65      1  

[5 rows x 22 columns]

In [3]:
#Checking unique values
df_churndata_modded.nunique()
# list(df_churndata_modded)

#There are many columns where there are 3 options
#Only a few columns: CustomerID, Tenure, MonthlyCharges, TotalCharges

Unnamed: 0          7043
customerID          7043
gender                 2
SeniorCitizen          2
Partner                2
Dependents             2
tenure                73
PhoneService           2
MultipleLines          3
InternetService        3
OnlineSecurity         3
OnlineBackup           3
DeviceProtection       3
TechSupport            3
StreamingTV            3
StreamingMovies        3
Contract               3
PaperlessBilling       2
PaymentMethod          4
MonthlyCharges      1585
TotalCharges        6531
Churn                  2
dtype: int64

In [4]:
#Checking any duplicates in CustomerID

# DataFrame.duplicated(subset=None, keep='first')

df_churndata_modded.duplicated(subset="customerID", keep='first')

len(df_churndata_modded.duplicated())

7043

### We see from below plot, there are alot of categorical values. Lots of straight lines. We see vertical lines for discrete values instead of a homogenous "cloud"

In [ ]:
#draw scatterplot matrix using plt

pd.plotting.scatter_matrix(df_churndata_modded, figsize=(21,21))
plt.show()

### We will make some scatterplots to compare different features

In [6]:
# sns.scatterplot(data=tips, x="total_bill", y="tip")

# sns.scatterplot(data=df_churndata_modded, x="TotalCharges", y="MonthlyCharges")

In [7]:
#Decision Tree! Jumping the gun here!
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier 
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import OneHotEncoder
from sklearn import tree

In [8]:
#Splitting the data

X = df_churndata_modded[['gender','SeniorCitizen','Partner','Dependents','PhoneService','MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling', 'PaymentMethod']]
# X = df_churndata_modded[['gender','Partner','Dependents','PhoneService','MultipleLines', 'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport','Contract']]
y = df_churndata_modded[['Churn']]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42)

In [9]:
#OHE

ohe = OneHotEncoder()

ohe.fit(X_train)
X_train_ohe = ohe.transform(X_train).toarray()

# Creating this DataFrame is not necessary its only to show the result of the ohe
ohe_df = pd.DataFrame(X_train_ohe, columns=ohe.get_feature_names(X_train.columns))

ohe_df.head()

gender_Female  gender_Male  SeniorCitizen_0  SeniorCitizen_1  Partner_No  \
0            0.0          1.0              1.0              0.0         0.0   
1            0.0          1.0              1.0              0.0         1.0   
2            0.0          1.0              1.0              0.0         1.0   
3            1.0          0.0              0.0              1.0         0.0   
4            1.0          0.0              0.0              1.0         1.0   

   Partner_Yes  Dependents_No  Dependents_Yes  PhoneService_No  \
0          1.0            0.0             1.0              0.0   
1          0.0            1.0             0.0              0.0   
2          0.0            1.0             0.0              0.0   
3          1.0            0.0             1.0              0.0   
4          0.0            1.0             0.0              0.0   

   PhoneService_Yes  ...  StreamingMovies_Yes  Contract_Month-to-month  \
0               1.0  ...                  0.0                      0.0   
1               1.0  ...                  0.0                      1.0   
2               1.0  ...                  0.0                      0.0   
3               1.0  ...                  1.0                      0.0   
4               1.0  ...                  0.0                      1.0   

   Contract_One year  Contract_Two year  PaperlessBilling_No  \
0                1.0                0.0                  0.0   
1                0.0                0.0                  0.0   
2                1.0                0.0                  1.0   
3                0.0                1.0                  0.0   
4                0.0                0.0                  0.0   

   PaperlessBilling_Yes  PaymentMethod_Bank transfer (automatic)  \
0                   1.0                                      0.0   
1                   1.0                                      0.0   
2                   0.0                                      0.0   
3                   1.0                                      0.0   
4                   1.0                                      0.0   

   PaymentMethod_Credit card (automatic)  PaymentMethod_Electronic check  \
0                                    0.0                             1.0   
1                                    0.0                             1.0   
2                                    1.0                             0.0   
3                                    0.0                             1.0   
4                                    0.0                             1.0   

   PaymentMethod_Mailed check  
0                         0.0  
1                         0.0  
2                         0.0  
3                         0.0  
4                         0.0  

[5 rows x 43 columns]

#training decision tree

# Create the classifier, fit it on the training data and make predictions on the test set
clf = DecisionTreeClassifier(criterion='entropy')

clf.fit(X_train_ohe, y_train)

In [ ]:
fig, axes = plt.subplots(nrows = 1,ncols = 1, figsize = (3,3), dpi=300)
tree.plot_tree(clf,
               feature_names = ohe_df.columns, 
               class_names=np.unique(y).astype('str'),
               filled = True)
plt.show()

#Whateven the hell am i looking at right now

In [12]:
#Honestly I did not need to see that crazy ass tree. Gonna test for accuracy

X_test_ohe = ohe.transform(X_test)
y_preds = clf.predict(X_test_ohe)

print('Accuracy: ', accuracy_score(y_test, y_preds))

Accuracy:  0.7141504969238051


## Oh, it's pruning time

### First I want to check if there are columns with same values that I can get rid of. Turns out there is differences in StreamingMovies and StreamingTV

### I use sum(x) because it will sum up all the true values which I got 5484 which is way off from 7043

In [20]:


#First let's compare streaming with the other streaming column
#StreamingTV StreamingMovies
# df_churndata_modded['StreamingTV']
# df_churndata_modded['StreamingMovies']

df_churndata_modded['StreamingMovies'].equals(df_churndata_modded['StreamingTV'])

# type(df_churndata_modded['StreamingMovies'])
# type(df_churndata_modded['StreamingTV'])


False

In [44]:
compare_Streaming = np.where(df_churndata_modded['StreamingMovies'] == df_churndata_modded['StreamingTV'], True, False)
x = compare_Streaming.tolist()
# print(np.all(compare_Streaming == False))
x
sum(x)



5484

In [39]:
df_churndata_modded['StreamingMovies'][0:8]

0     No
1     No
2     No
3     No
4     No
5    Yes
6     No
7     No
Name: StreamingMovies, dtype: object

In [40]:
df_churndata_modded['StreamingTV'][0:8]

0     No
1     No
2     No
3     No
4     No
5    Yes
6    Yes
7     No
Name: StreamingTV, dtype: object

In [ ]:
#So should I combine StreamingTV and StreamingMovies? 
